components/brave_wallet/browser/internal/hd_key_ed25519.cc

```diff
@@ -62,74 +87,83 @@ std::string HDKeyEd25519::GetPath() const {
-    return nullptr;
-  }
+  std::array<uint8_t, 37> hmac_payload = {};
+
+  auto span_writer = base::SpanWriter(base::span(hmac_payload));
+  // https://github.com/satoshilabs/slips/blob/master/slip-0010.md#private-parent-key--private-child-key
+  span_writer.Write(base::U8ToBigEndian(0));
+  span_writer.Write(GetPrivateKeyAsSpan());
+  span_writer.Write(base::U32ToBigEndian(index));
+  DCHECK_EQ(span_writer.remaining(), 0u);
 
-  return std::make_unique<HDKeyEd25519>(path, std::move(child_private_key));
+  return base::WrapUnique(new HDKeyEd25519(chain_code_.AsSpan(), hmac_payload));
 }
 
-std::vector<uint8_t> HDKeyEd25519::Sign(base::span<const uint8_t> msg) {
-  auto signature_result = private_key_->unwrap().sign(
-      rust::Slice<const uint8_t>{msg.data(), msg.size()});
-  if (!signature_result->is_ok()) {
-    VLOG(0) << std::string(signature_result->error_message());
-    return std::vector<uint8_t>();
+// static
+std::unique_ptr<HDKeyEd25519> HDKeyEd25519::GenerateFromSeedAndPath(
+    base::span<const uint8_t> seed,
+    std::string_view hd_path) {
+  auto hd_key = base::WrapUnique(
+      new HDKeyEd25519(base::byte_span_from_cstring(kMasterSecret), seed));
+
+  auto nodes = ParseFullHDPath(hd_path);
+  if (!nodes) {
+    return nullptr;
   }
-  auto signature_bytes = signature_result->unwrap().to_bytes();
-  return std::vector<uint8_t>(signature_bytes.begin(), signature_bytes.end());
+
+  for (auto index : *nodes) {
+    if (index < kHardenedOffset) {
+      return nullptr;
+    }
+    hd_key = hd_key->DeriveChild(index);
+    if (!hd_key) {
+      return nullptr;
+    }
+  }
+
+  return hd_key;
 }
 
-bool HDKeyEd25519::VerifyForTesting(base::span<const uint8_t> msg,
-                                    base::span<const uint8_t> sig) {
-  auto verification_result = private_key_->unwrap().verify(
-      rust::Slice<const uint8_t>{msg.data(), msg.size()},
-      rust::Slice<const uint8_t>{sig.data(), sig.size()});
-  if (!verification_result->is_ok()) {
-    VLOG(0) << std::string(verification_result->error_message());
-    return false;
+std::optional<std::array<uint8_t, kEd25519SignatureSize>> HDKeyEd25519::Sign(
+    base::span<const uint8_t> msg) {
+  // For the sake of performance `ED25519_sign` does not derive public key from
+  // private key, but assumes they are both already available as key_pair.
+  std::array<uint8_t, kEd25519KeypairSize> key_pair =
+      MakeKeyPair(private_key_.AsSpan(), public_key_);
+
+  std::array<uint8_t, kEd25519SignatureSize> signature = {};
+  if (!ED25519_sign(signature.data(), msg.data(), msg.size(),
+                    key_pair.data())) {
+    return std::nullopt;
   }
-  return true;
+  return signature;
 }
 
-std::vector<uint8_t> HDKeyEd25519::GetPrivateKeyBytes() const {
-  auto secret_key = private_key_->unwrap().secret_key_raw();
-  return {secret_key.begin(), secret_key.end()};
+base::span<const uint8_t, kEd25519PrivateKeySize>
```

Comment:
are these changes necessary for something in the new implementation? Please don't change anything that doesn't strictly need to be changed. Any other changes should be follow-ups

Reply:
agreed